### ISSUE TITLE
Save logs on stop/crash

### ISSUE THREAD

**Issue by stuartpb** [plushu/plushu-logs#1]
If we can do this, that would be fantastic.

Ideas:
- systemd/upstart unit to do it on shutdown
- Hookify app shutdown during restart to make this saveable on redeploy

**Comment by stuartpb** [comment]
We can also have nginx logs go to a separate set of files

**Comment by stuartpb** [comment]
Is there something we can set in docker to make a detached container shoot a program's stdout/stderr to a file in realtime?

**Comment by stuartpb** [comment]
We should really save logs any time we're going to delete a cidfile, that should be part of a hook (Above the actual "docker stop"

**Comment by stuartpb** [comment]
Okay, so we have two sets of logs:
- Archived logs from older containers
- Logs from the current container

These will be implemented in two separate plugins, with the "current container" logs being saved right before the container is destroyed.

As for stopped/crashed containers, Docker saves those logs, so we have them as long as the container stays around (assuming nothing on the system does a cleanup).

All logs will take a format of "[timestamp] [message]" for each line (assuming Docker handles multi-line messages properly): we can figure out per-source insertion later.

There will then be an app-logs hook, which these two plugins will output log output for. Initially, they'll just output everything in sensible order (archives come before live logs), but later in development, the caller will be in charge of sorting (and maybe filtering and/or coloring) their output.

(App logs for lifecycle will come later, in their own plugins.)

**Comment by stuartpb** [comment]
Y'know, it kind of feels like the actual _archiving_ of the logs is going to be in the same plugin responsible for _displaying_ the logs, but _not_ the plugin responsible for outputting the logs once archived... which is weird, but correct.

**Comment by stuartpb** [comment]
In fact, now that I think about it, I'm starting to think plugins should handle their own log output (so saving the output of the old container is a job for deploy-app-local-container). Here's why:
- The log should only get saved if the cleanup works properly - if the container won't clean up for some reason, we don't want a duplicate of its log entries added to the log every time.
- Build and release (#6) control where the output from each of those goes, so they need to handle logging anyway.

**Comment by stuartpb** [comment]
In _fact_, we want to do the log saving _between when the container is stopped and when we remove it_, so we're going to throw a `docker stop` and the log saving in front of `docker-rm-cidfile`.

(In fact, now that I think of it, it's probably enough to append the log after stopping the container. Any errors that manifest in removing the container _after_ cleanly stopping it are going to require some fairly catastrophic cleanup anyway.)

**Comment by stuartpb** [comment]
Okay, after a little more thought, these plugins will offer the logs up to archive hooks (which may or may not be listening). plushu-archive-app-logs will handle the archiving, and retrieval from archive, of these logs.

**Comment by stuartpb** [comment]
Implemented, tested, and fixed.

**Comment by stuartpb** [comment]
**tl;dr** we're letting Docker "save" the logs until we destroy the container.
